rust-wrapper/src/msm.rs

```diff
@@ -0,0 +1,77 @@
+use ark_bls12_381::{Fr as ScalarField, G1Affine as GAffine, G1Projective as G};
+use ark_ec::VariableBaseMSM as BaselineVariableBaseMSM;
+use ark_ff::PrimeField;
+use ark_msm::msm::VariableBaseMSM;
+use ark_serialize::CanonicalSerialize;
+use ark_std::log2;
+use haskell_ffi::{from_haskell::marshall_from_haskell_var, to_haskell::marshall_to_haskell_var};
+use utils::{Buffer, RW};
+
+use crate::utils::{self, deserialize_vector_points, deserialize_vector_scalar_field};
+
+#[no_mangle]
+pub extern "C" fn rust_wrapper_multi_scalar_multiplication(
+    points_var: *const u8,
+    points_len: usize,
+    scalars_var: *const u8,
+    scalars_len: usize,
+    out: *mut u8,
+    out_len: &mut usize,
+) {
+    let scalars =
+        marshall_from_haskell_var::<RW, Buffer<ScalarField>>(scalars_var, scalars_len, RW).0;
+    let points = marshall_from_haskell_var::<RW, Buffer<GAffine>>(points_var, points_len, RW).0;
+
+    let r: GAffine = G::msm(&points, &scalars).unwrap().into();
+
+    marshall_to_haskell_var(&r, out, out_len, RW);
+}
+// This is inner function from arkmsm crate, derived from benchmark results. May not be optimal for all configurations
+// https://github.com/snarkify/arkmsm/blob/main/src/msm.rs
+const fn get_opt_window_size(k: u32) -> u32 {
+    match k {
+        0..=9 => 8,
+        10..=12 => 10,
+        13..=14 => 12,
+        15..=19 => 13,
+        20..=22 => 15,
+        23.. => 16,
+    }
+}
+
+///
+/// # Safety
+/// The caller must ensure that valid pointers and sizes are passed. 
+/// To find out how data should be represented in memory, you can look at msm_bench.rs
+/// .
+#[no_mangle]
+pub unsafe extern "C" fn rust_wrapper_multi_scalar_multiplication_without_serialization(
```

Comment:
The most part of the function is safe, once again it is better to extract the safe part into separate safe function and leave the unsafe part in FFI